pyqt-apps/siriushla/as_ap_measure/emittance_meas_lnls.py

```diff
@@ -507,3 +524,391 @@ def pb_stop_clicked(self):
         """
         _log.info('Stopping...')
         self._measuring.set()
+
+
+def gettransmat(elem, L, K1=None, B=None):
+    R = np.eye(4)
+
+    if elem.lower().startswith('qu') and K1 is not None and K1 == 0:
+        elem = 'drift'
+    if elem.lower().startswith('dr'):
+        R = np.array([
+            [1, L, 0, 0],
+            [0, 1, 0, 0],
+            [0, 0, 1, L],
+            [0, 0, 0, 1],
+        ])
+    elif elem.lower().startswith('qu') and K1 is not None:
+        kq = np.sqrt(abs(K1))
+        c = np.cos(kq*L)
+        s = np.sin(kq*L)
+        ch = np.cosh(kq*L)
+        sh = np.sinh(kq*L)
+        if K1 > 0:
+            x11, x12, x21 = c,  1/kq*s, -kq*s
+            y11, y12, y21 = ch, 1/kq*sh, kq*sh
+        else:
+            x11, x12, x21 = ch, 1/kq*sh, kq*sh
+            y11, y12, y21 = c,  1/kq*s, -kq*s
+        R = np.array([
+            [x11, x12, 0,   0],
+            [x21, x11, 0,   0],
+            [0,   0,   y11, y12],
+            [0,   0,   y21, y11],
+        ])
+    return R
+
+
+def _calc_moments(axis, proj):
+    dx = axis[1]-axis[0]
+    Norm = np.trapz(proj, dx=dx)
+    cen = np.trapz(proj*axis, dx=dx)/Norm
+    sec = np.trapz(proj*axis*axis, dx=dx)/Norm
+    std = np.sqrt(sec - cen*cen)
+    return cen, std
+
+
+def _gaussian(x, amp, mu, sigma, y0):
+    return amp*np.exp(-(x-mu)**2.0/(2.0*sigma**2.0))+y0
+
+
+def _fit_gaussian(x, y, amp=None, mu=None, sigma=None, y0=None):
+    amp = amp or np.amax(y)
+    par = _calc_moments(x, y)
+    mu = mu or par[0]
+    sigma = sigma or par[1]
+    y0 = y0 or np.mean(y)
+    try:
+        p_opt, p_cov = curve_fit(_gaussian, x, y, (amp, mu, sigma, y0))
+    except Exception:
+        p_opt = (amp, mu, sigma, y0)
+        print('Fitting Problem')
+    return p_opt
+
+
+class MatplotlibWidget(Canvas):
```

Comment:
This re-implementation could be moved to siriushla.widgets.figure_canvas.SiriusFigureCanvas.

Reply:
done!